crates/gitbutler-branch-actions/src/status.rs

```diff
@@ -27,21 +27,39 @@ pub struct VirtualBranchesStatus {
     pub skipped_files: Vec<gitbutler_diff::FileDiff>,
 }
 
+pub fn get_applied_status(
+    ctx: &CommandContext,
+    perm: Option<&mut WorktreeWritePermission>,
+) -> Result<VirtualBranchesStatus> {
+    get_applied_status_cached(ctx, perm, None)
+}
+
 /// Returns branches and their associated file changes, in addition to a list
 /// of skipped files.
+/// `worktree_changes` are all changed files against the current `HEAD^{tree}` and index
+/// against the current working tree directory, and it's used to avoid double-computing
+/// this expensive information.
 // TODO(kv): make this side effect free
-pub fn get_applied_status(
+#[instrument(level = tracing::Level::DEBUG, skip(ctx, perm, worktree_changes))]
+pub fn get_applied_status_cached(
     ctx: &CommandContext,
     perm: Option<&mut WorktreeWritePermission>,
+    worktree_changes: Option<gitbutler_diff::DiffByPathMap>,
 ) -> Result<VirtualBranchesStatus> {
     assure_open_workspace_mode(ctx).context("ng applied status requires open workspace mode")?;
-    let integration_commit = get_workspace_head(ctx)?;
+    // TODO(ST): this was `get_workspace_head()`, which is slow and ideally, we don't dynamically
+    //           calculate which should already be 'fixed' - why do we have the integration branch
+    //           if we can't assume it's in the right state? So ideally, we assure that the code
+    //           that affects the integration branch also updates it?
+    let integration_commit_id = ctx.repository().head_commit()?.id();
```

Comment:
@Caleb-T-Owens @krlvi This change was made to assure that the passed-in `worktree_changes` wouldn't differ with what's produced here, as in `get_uncommitted_files()` the integration-commit would indeed be `HEAD`, not something freshly computed.

However, some tests relied on this behaviour, and I fixed them by forcing an integration-commit update.
The question is though if the application also assumes that `list_virtual_branches()` (which calls this function) picks up the latest changes.

My expectation would be that places that alter the state and affect the integration commit, will also update it, and I hope it's realistic.

Doing this is also in the interest of performance, as `get_workspace_head()` is quite expensive to call.

Reply:
It looks like E2E has one failing test due to that, will look more closely tomorrow. This might be the point where the app has to adapt to get an update-to-date worktree commit after a relevant change is performed.

Reply:
It was flakiness, that's good to know.

Reply:
reading this code im still scratching my head if we ever needed it. Lets do some extensive testing with the nightly to know this works correctly still